rustls/src/crypto/hpke.rs

```diff
@@ -0,0 +1,131 @@
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::fmt::Debug;
+
+use crate::msgs::enums::{HpkeAead, HpkeKdf, HpkeKem};
+use crate::Error;
+
+/// A provider for [RFC 9180] Hybrid Public Key Encryption (HPKE).
+///
+/// At a minimum each provider must support base mode and the [HPKE ciphersuite profile] required for
+/// encrypted client hello (ECH):
+///  * KEM: DHKEM(X25519, HKDF-SHA256)
+///  * KDF: HKDF-SHA256
+///  * AEAD: AES-128-GCM
+///
+/// [RFC 9180]: <https://www.rfc-editor.org/rfc/rfc9180.html>
+/// [HPKE ciphersuite profile]: <https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-9>
+pub trait HpkeProvider: Debug + Send + Sync + 'static {
+    /// Create an HPKE instance for the given mode, KEM, KDF and AEAD.
+    ///
+    /// May return an error if the mode, KEM, KDF or AEAD are unsupported by the HPKE provider.
+    fn setup(
+        &self,
+        mode: Mode,
+        kem: HpkeKem,
+        kdf: HpkeKdf,
+        aead: HpkeAead,
+    ) -> Result<Box<dyn Hpke>, Error>;
+
+    /// Does the provider support the given mode?
+    fn supports_mode(&self, mode: Mode) -> bool;
+
+    /// Does the provider support the given [HpkeKem]?
+    fn supports_kem(&self, kem: HpkeKem) -> bool;
+
+    /// Does the provider support the given [HpkeKdf]?
+    fn supports_kdf(&self, kdf: HpkeKdf) -> bool;
+
+    /// Does the provider support the given [HpkeAead]?
+    fn supports_aead(&self, aead: HpkeAead) -> bool;
+}
+
+/// Available HPKE authentication modes.
+///
+/// The base case for HPKE is unauthenticated encryption to a public key. In addition to this
+/// mode there are three modes to provide a form of authentication.
+///
+/// See [RFC 9180 Section 5](https://www.rfc-editor.org/rfc/rfc9180.html#name-hybrid-public-key-encryptio)
+/// for more details.
+#[derive(PartialEq, Copy, Clone, Debug)]
+#[repr(u8)]
+pub enum Mode {
+    /// Encryption to a public key, no additional authentication.
+    Base = 0x00,
+
+    /// Encryption to a public key, authenticated with a pre-shared secret key.
+    Psk = 0x01,
+
+    /// Encryption to a public key, authenticated with the sender's private key.
+    Auth = 0x02,
+
+    /// Encryption to a public key, authenticated with the sender's private key and a
+    /// pre-shared secret key.
+    AuthPsk = 0x03,
+}
+
+/// An HPKE instance that can produce sender and receiver contexts.
+pub trait Hpke: Debug + Send + Sync {
+    /// Create a sender context to encrypt messages to a receiver.
+    ///
+    /// This will return an encapsulated secret and a sender context. The sender context
+    /// can be used to seal messages to the receiver.
+    fn setup_sender(
+        &self,
+        pk_r: &HpkePublicKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
```

Comment:
Should we wrap up the arguments shared by `setup_sender()` and `setup_receiver()` in a common type?

Reply:
We could wrap the `Optional` ones into something describing authentication, but per [my other comment](https://github.com/rustls/rustls/pull/1589#issuecomment-1810932503) maybe we'd prefer to drop them entirely and focus only on base mode?

Reply:
This fell away with the change to focus on base mode.